### ISSUE TITLE
Yoga federation driver support

### ISSUE THREAD

**Issue by Fedepnt** [MichalLytek/typegraphql-nestjs#52]
Hi @MichalLytek, I'm writing you because I'm currently splitting my monolith graph into microservices but I encountered an issue.
My stack is composed by Nestjs (v10.0.0) + typegraphql-nestjs (v0.6.0) + Yoga GraphQL (v5.0.0).

To achieve the migration to microservices I'm going to use the Federation V2 with the [YogaFederationDriver](https://the-guild.dev/graphql/yoga-server/docs/integrations/integration-with-nestjs#apollo-federation). 
Here's my main module after the migration:
 ```
import * as path from 'path';
import { Module } from '@nestjs/common';
import { ConfigModule } from '@nestjs/config';
import { MongooseModule } from '@nestjs/mongoose';
import { YogaFederationDriver } from '@graphql-yoga/nestjs-federation';
import { TypeGraphQLModule } from 'typegraphql-nestjs';
import { UserModule } from './user/user.module';
import { SubGraphRequest } from './dto/types/request.type';

 @Module({
  imports: [
    ConfigModule.forRoot(),
    MongooseModule.forRoot(process.env.MONGODB_URL, {
      maxPoolSize: 10,
    }),

    TypeGraphQLModule.forRootAsync({
      driver: YogaFederationDriver,
      useFactory: () => ({
        federationVersion: 2,
        emitSchemaFile: path.join(__dirname, '..', 'schema.graphQL'),
        context: async ({ req }: { req: SubGraphRequest }) => {
          return {
            email: req.get('x-user-email')
          };
        }
      }),
    }),
    
    // my modules 
    UserModule,
  ],
})
export class AppModule {}
```
Unfortunately this generates the graphql schema error:

```
/user-subgraph/node_modules/@apollo/federation-internals/src/error.ts:36
  err: (message: string, options?: GraphQLErrorOptions) => new GraphQLError(
                                                           ^
GraphQLError: Type User must define one or more fields.
```

After a deep investigation I found the issue is caused by the following line and I noticed that for the federation part the supported driver is `ApolloFederationDriver` from `@nestjs/apollo`
https://github.com/MichalLytek/typegraphql-nestjs/blob/8f9bc06e443321b0187437f7d1aef7ce17bd92c5/src/typegraphql-options.factory.ts#L39
I tried to change this line with this one
`const isFederatedModule = driver === ApolloFederationDriver || driver === YogaFederatedDrivers;` 
and it seems to work. 

Is this the expected behaviour? Is it possible to extend the support to Yoga Federated Drivers and the other available drivers?

And also maybe it could be useful to check the federation version instead of the drivers types, for example
`const isFederatedModule = federationVersion === 1 | federationVersion === 2;`

**Comment by MichalLytek** [comment]
Yeah, you're right 😉 Can you check in NestJS source code how do they handle the differentiation?

**Comment by Fedepnt** [comment]
Seems they are using the federation version param from the `.forRootAsync` config [AutoSchemaFileValue](https://github.com/nestjs/graphql/blob/8f496d4f2cb3b1173c56e8c2daec3bdb45edd513/packages/graphql/lib/interfaces/schema-file-config.interface.ts#L37)

```
GraphQLModule.forRootAsync<YogaFederationDriverConfig>({
      driver: YogaFederationDriver,
      useFactory: () => {
        return {
          autoSchemaFile: {
            federation: 2,
          },
          context: async ({ req }: { req: SubGraphRequest }) => {
            return {
              email: req.get('x-user-email'),
            };
          }
        }
      },
    }),
```
and they are using it to generate the schema later on [GraphQLFederationFactory](https://github.com/nestjs/graphql/blob/8f496d4f2cb3b1173c56e8c2daec3bdb45edd513/packages/graphql/lib/federation/graphql-federation.factory.ts#L358)
```
private getFederationVersionAndConfig(
    autoSchemaFile: AutoSchemaFileValue,
  ): [FederationVersion, FederationConfig?] {
    if (!autoSchemaFile || typeof autoSchemaFile !== 'object') {
      return [DEFAULT_FEDERATION_VERSION];
    }
    if (typeof autoSchemaFile.federation !== 'object') {
      return [autoSchemaFile.federation ?? DEFAULT_FEDERATION_VERSION];
    }
    return [
      autoSchemaFile.federation?.version ?? DEFAULT_FEDERATION_VERSION,
      autoSchemaFile.federation,
    ];
  }
```
and

```
const [federationVersion, federationOptions] =
      this.getFederationVersionAndConfig(options.autoSchemaFile);

    const typeDefsDecorator = this.typeDefsDecoratorFactory.create(
      federationVersion,
      apolloSubgraphMajorVersion,
    );
    if (typeDefsDecorator) {
      typeDefs = typeDefsDecorator.decorate(typeDefs, federationOptions);
    }

    let executableSchema: GraphQLSchema = buildFederatedSchema({
      typeDefs: gql(typeDefs),
      resolvers: this.getResolvers(options.resolvers),
    });

    executableSchema = this.overrideOrExtendResolvers(
      executableSchema,
      autoGeneratedSchema,
      printSubgraphSchema,
    );

    const schema = options.schema
      ? mergeSchemas({
          schemas: [options.schema, executableSchema],
        })
      : executableSchema;
    return schema;
```

**Comment by MichalLytek** [comment]
@Fedepnt Implemented in 148bc99 🔒
